app/models/metrics/base_metric.rb

```diff
@@ -84,23 +90,28 @@ def evaluate(var)
 
     def calculate_all_values
       if values.nil? || timestamp.nil? || timestamp + interval < Time.zone.now
-        @values = {}.tap do |v|
-          rows.each do |row|
-            if respond_to?(:calculate_row)
-              # do calculation a row at a time
-              row_values = ActiveSupport::HashWithIndifferentAccess.new(calculate_row(row))
-              row_values.default = 0
-              columns.each do |column|
-                v[value_key(column, row)] = row_values[column]
-              end
-              # sleep for a very short while to avoid overloading the system
-              sleep(rand(0..0.1).round(2))
-            else
-              # do calculation a cell at a time
-              columns.each do |column|
-                v[value_key(column, row)] = calculate(column, row)
+        if respond_to?(:calculate_table)
+          @values = calculate_table
+          pause
+        else
+          @values = {}.tap do |v|
```

Comment:
How about moving this inner block of code to a new method named `calculate_individual_rows` - it doesn't look like it would need any additional parameters to be passed in so would be easy to refactor. The main benefit of that is that it makes it much easier to see the overall high level logic within the `calculate_all_values` method at a glance.

Reply:
now separated into a `calculate_individual_rows` method